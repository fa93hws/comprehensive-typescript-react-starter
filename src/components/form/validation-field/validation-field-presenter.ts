import { action, observable } from 'mobx';

export class FieldStore {
  @observable.ref
  public value: string = '';

  @observable.ref
  public errorMessage: string | null = null;
}

export class FieldPresenter {
  @action
  public static setValue(store: FieldStore, val: string) {
    store.value = val;
  }

  @action
  public static setError(store: FieldStore, errorMessage: string | null) {
    store.errorMessage = errorMessage;
  }

  public static hasError(store: FieldStore) {
    return store.errorMessage != null;
  }
}
