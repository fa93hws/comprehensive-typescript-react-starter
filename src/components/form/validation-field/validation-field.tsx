import * as cn from 'classnames';
import * as React from 'react';
import styles from './validation-field.css';

export type FieldType = 'text' | 'password';

interface ValidatedFieldProps {
  title: string;
  type: FieldType;
  value: string;
  errorMessage: string | null;
  onChange(val: React.ChangeEvent<HTMLInputElement>): void;
  onBlur?(): void;
}

export const ValidationField = (props: ValidatedFieldProps) => (
  <div>
    <h4>
      {props.title}
    </h4>
    <input
      type={props.type}
      value={props.value}
      onBlur={props.onBlur}
      onChange={props.onChange}
      className={cn({
        [styles.error]: props.errorMessage != null,
      })}
    />
    <p>
      {props.errorMessage}
    </p>
  </div>
);
