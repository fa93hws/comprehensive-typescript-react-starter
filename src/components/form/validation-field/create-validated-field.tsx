import * as React from 'react';
import { observer } from 'mobx-react';
import { ValidationController } from '../../base/validation/validation-controller';
import { FieldStore, FieldPresenter } from './validation-field-presenter';
import { FieldType, ValidationField } from './validation-field';

interface ValidatedFieldConfig {
  title: string;
  validationController: ValidationController<string>;
  type: FieldType;
  onChange?(val: string): void;
}

export function createValidationField(config: ValidatedFieldConfig) {
  const store = new FieldStore();
  const doValidate = () => {
    const error = config.validationController.getNormalizedError(store.value);
    FieldPresenter.setError(store, error);
  };
  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const val = event.target.value;
    FieldPresenter.setValue(store, val);
    doValidate();
    config.onChange && config.onChange(val);
  };
  const onBlur = () => {
    config.validationController.startValidate();
    doValidate();
  };

  const Field = observer(() => (
    <ValidationField
      onBlur={onBlur}
      type={config.type}
      onChange={onChange}
      value={store.value}
      title={config.title}
      errorMessage={store.errorMessage}
    />
  ));

  return {
    Field,
    store,
    doValidate,
  };
}
