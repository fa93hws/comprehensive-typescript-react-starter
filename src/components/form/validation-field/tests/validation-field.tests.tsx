import * as React from 'react';
import { ValidationField } from '../validation-field';

describe('ValidationField', () => {
  it('should render correctly for empty value without error message', () => {
    expect(
      <ValidationField
        title="title"
        type="text"
        value=""
        errorMessage={null}
        onChange={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('should render correctly for empty value with error message', () => {
    expect(
      <ValidationField
        title="title"
        type="text"
        value=""
        errorMessage="there is an error"
        onChange={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('should render correctly for some value without error message', () => {
    expect(
      <ValidationField
        title="title"
        type="text"
        value="there is some value"
        errorMessage={null}
        onChange={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('should render correctly for some value with error message', () => {
    expect(
      <ValidationField
        title="title"
        type="text"
        value="there is some value"
        errorMessage="there is some errors"
        onChange={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });
});
