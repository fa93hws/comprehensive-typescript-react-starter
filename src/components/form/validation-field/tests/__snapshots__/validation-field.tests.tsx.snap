// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationField should render correctly for empty value with error message 1`] = `
<div>
  <h4>
    title
  </h4>
  <input
    class="undefined"
    type="text"
    value=""
  />
  <p>
    there is an error
  </p>
</div>
`;

exports[`ValidationField should render correctly for empty value without error message 1`] = `
<div>
  <h4>
    title
  </h4>
  <input
    class=""
    type="text"
    value=""
  />
  <p />
</div>
`;

exports[`ValidationField should render correctly for some value with error message 1`] = `
<div>
  <h4>
    title
  </h4>
  <input
    class="undefined"
    type="text"
    value="there is some value"
  />
  <p>
    there is some errors
  </p>
</div>
`;

exports[`ValidationField should render correctly for some value without error message 1`] = `
<div>
  <h4>
    title
  </h4>
  <input
    class=""
    type="text"
    value="there is some value"
  />
  <p />
</div>
`;
